components/brave_shields/browser/ad_block_subscription_service_manager.cc

```diff
@@ -0,0 +1,439 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_shields/browser/ad_block_subscription_service_manager.h"
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "base/files/file_util.h"
+#include "base/strings/string_util.h"
+#include "base/task/post_task.h"
+#include "base/values.h"
+#include "brave/browser/download/brave_download_service_factory.h"
+#include "brave/common/pref_names.h"
+#include "brave/components/adblock_rust_ffi/src/wrapper.h"
+#include "brave/components/brave_shields/browser/ad_block_service_helper.h"
+#include "brave/components/brave_shields/browser/ad_block_subscription_service.h"
+#include "brave/components/brave_shields/browser/ad_block_subscription_service_manager_observer.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/browser/profiles/profile_key.h"
+#include "chrome/browser/profiles/profile_manager.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "content/public/browser/browser_thread.h"
+
+namespace brave_shields {
+
+void AdBlockSubscriptionServiceManager::OnSystemProfileCreated(
+    Profile* profile,
+    Profile::CreateStatus status) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  DCHECK(profile->IsSystemProfile());
+  DCHECK_NE(status, Profile::CREATE_STATUS_LOCAL_FAIL);
+  if (status != Profile::CREATE_STATUS_INITIALIZED) {
+    return;
+  }
+
+  InitializeDownloadManager(profile);
+}
+
+// Force the system profile to be created. Without this call, it is eventually
+// lazy-loaded by other services using the same mechanism.
+void AdBlockSubscriptionServiceManager::InitializeSystemProfile() {
+  g_browser_process->profile_manager()->CreateProfileAsync(
```

Comment:
I really don't think we should ever do something like this

Reply:
Let's see what @bridiver has to say - I'm not a huge fan of it either but this is a solution we settled on together.

Reply:
We've already hit into a problem with eagerly initialized profile, I'm pretty sure this code can lead to much more unexpected consequences

Reply:
@iefremov this is actually used everywhere the system profile is needed. It is always lazy loaded.

Reply:
But why do we need to instantiate the system profile to download files from the web?.. I mean, even component updater doesn't do that. Are we solving a more complex task than the component updater?